onnxscript/rewriter/ort_fusions/gqa.py

```diff
@@ -2,156 +2,282 @@
 
-        return query, key, value
+    total_seq_len_0D = op.Add(past_seq_len_0D, seq_len_0D)
+    total_seq_len = op.Reshape(total_seq_len_0D, [-1])
+
+    # The Phi modeling code generates the following +1 as the target-length, which seems
+    # unnecessary in this context. But using it for pattern-matching against
+    # generated onnx model.
+    total_seq_len_plus_1_0D = op.Add(total_seq_len_0D, 1)
+    total_seq_len_plus_1 = op.Reshape(total_seq_len_plus_1_0D, [-1])
+
+    current_range = op.Range(past_seq_len_0D, total_seq_len_0D, 1)
+    mask_shape = op.Concat(seq_len, total_seq_len_plus_1, axis=0)
+    mask_all_min = op.Expand(-3.4028235e38, mask_shape)
+    total_range_as_row = op.Range(0, total_seq_len_plus_1_0D, 1)
+    current_range_as_column = op.Reshape(current_range, [-1, 1])
+    boolean_mask = op.Greater(total_range_as_row, current_range_as_column)
+    float_0_1_mask = op.Cast(boolean_mask, to=1)
+    float_0_min_mask = op.Mul(mask_all_min, float_0_1_mask)
+    mask_4d = op.Unsqueeze(float_0_min_mask, [0, 1])
+    mask_B1ST_plus = op.Expand(mask_4d, shape_B111)
+
+    # Get rid of the extra +1 added above: total_seq_len is enough, no
+    # need for total_seq_len+1.
+    mask_B1ST = op.Slice(mask_B1ST_plus, [0], total_seq_len, [3], [1])
+    return mask_B1ST
+
+
+class GroupQueryAttention(pattern.RewriteRuleClassBase):
+    def __init__(self):
+        super().__init__("GQA", remove_nodes=False)
 
     def pattern(
         self,
         op,
-        input,
-        qkv_weight,
-        mask,
-        cos,
-        sin,
+        query_BSD,
+        key_BSDkv,
+        value_BSDkv,
         past_key,
         past_value,
-        position_ids,
+        input_ids,
+        past_seq_length,
+        total_seq_length,
+        cos,
+        sin,
+        some_kv_cache,
+        shape_B111,
     ):
-        query, key, value = self._compute_packed_QKV(op, input, qkv_weight)
+        # Reshape query from (B, S, D) to (B, S, H, D/H)
+        query_BSHDh = op.Reshape(query_BSD, _allow_other_inputs=True, _outputs=["query_BSHDh"])
+        # Transpose from (B, S, H, D/H) to (B, H, S, D/H)
+        query_BHSDh = op.Transpose(query_BSHDh, perm=[0, 2, 1, 3])
+
+        # Reshape key from (B, S, Dkv) to (B, S, Hkv, D/H)
+        key_BSHkvDh = op.Reshape(key_BSDkv, _allow_other_inputs=True, _outputs=["key_BSHkvDh"])
+        # Transpose from (B, S, Hkv, D/H) to (B, Hkv, S, D/H)
+        key_BHkvSDh = op.Transpose(key_BSHkvDh, perm=[0, 2, 1, 3])
+
+        # Reshape value from (B, S, Dkv) to (B, S, Hkv, D/H)
+        value_BSHkvDh = op.Reshape(
+            value_BSDkv, _allow_other_inputs=True, _outputs=["value_BSHkvDh"]
+        )
+        # Transpose from (B, S, Hkv, D/H) to (B, Hkv, S, D/H)
+        value_BHkvSDh = op.Transpose(value_BSHkvDh, perm=[0, 2, 1, 3])
 
-        query_rope = op.RotaryEmbedding(query, position_ids, cos, sin, _domain="com.microsoft")
+        position_ids = op.Range(past_seq_length, total_seq_length, 1)
```

Comment:
Do we plan to support variations where position_ids is provided as an input?

Reply:
Yes. But these are currently driven by specific models. We need a model with the other flavor to develop that variant.